File: src/com/patterns/dynamicProgramming/easy/python/rod_cutting.py
# --------------------------------- TOP DOWN DP ---------------------------------

def rod_cutting_top_down_dp(price, length, max_profits_table, cuts_table):

    if length == 0:

        return 0, cuts_table

    if max_profits_table[length] == -1:

        for i in range(1, length + 1):

            price_if_cut_at_length_i = price[i] + rod_cutting_top_down_dp(price, length - i,
                                                                          max_profits_table, cuts_table)[0]

            if price_if_cut_at_length_i > max_profits_table[length]:

                max_profits_table[length] = price_if_cut_at_length_i
                cuts_table[length] = i

    return max_profits_table[length], cuts_table


def top_down_driver_function(price, length):

    max_profits_table = [-1 for _ in range(length + 1)]
    cuts_table = [0 for _ in range(length + 1)]

    max_profit, cuts = rod_cutting_top_down_dp(price, length, max_profits_table, cuts_table)

    n = length
    answer_table = []

    while n > 0:

        answer_table.append(cuts_table[n])
        n -= cuts_table[n]

    return answer_table, max_profit


# --------------------------------- BOTTOM UP DP ---------------------------------


def rod_cutting_bottom_up_dp(price, length):

    max_profit_table = [0 for _ in range(length + 1)]
    cuts_table = [0 for _ in range(length + 1)]

    for i in range(1, length + 1):

        for j in range(1, i + 1):

            price_if_cut_at_length_j_of_rod_with_length_i = price[j] + max_profit_table[i - j]

            if price_if_cut_at_length_j_of_rod_with_length_i > max_profit_table[i]:

                max_profit_table[i] = price_if_cut_at_length_j_of_rod_with_length_i
                cuts_table[i] = j

    n = length
    answer_table = []

    while n > 0:

        answer_table.append(cuts_table[n])
        n -= cuts_table[n]

    return answer_table, max_profit_table[length]


# --------------------------------- TEST CODE ---------------------------------

price_arr = [0, 2, 5, 9, 10, 12, 13, 15, 16]
rod_length = 8

print(top_down_driver_function(price_arr, rod_length))
print(rod_cutting_bottom_up_dp(price_arr, rod_length))

File: src/com/patterns/dynamicProgramming/easy/java/RodCutting.java
package com.patterns.dynamicProgramming.easy.java;

import java.util.ArrayList;
import java.util.List;

public class RodCutting {

    public static void main(String[] args) {

        int rodLength = 8;
        int[] testPrices = {0, 2, 5, 9, 10, 12, 13, 15, 16};

        System.out.println(rodCuttingDriver(rodLength, testPrices));
        System.out.println(rodCuttingBottomUpDp(rodLength, testPrices));

    }

    private static RodCuttingAnswerObject rodCuttingDriver(int len, int[] prices) {

        int[] maxProfitsTable = new int[len + 1];
        int[] cutsTable = new int[len + 1];

        RodCuttingReturnObject answer = rodCuttingTopDownDp(len, prices, maxProfitsTable, cutsTable);

        List<Integer> answerTable = new ArrayList<>();

        int n = len;

        while (n > 0) {
            answerTable.add(answer.getCutsTable()[n]);
            n -= answer.getCutsTable()[n];
        }

        return new RodCuttingAnswerObject(answer.getProfit(), answerTable);

    }

    private static RodCuttingReturnObject rodCuttingTopDownDp(int len, int[] prices, int[] maxProfitsTable, int[] cutsTable) {

        if (len == 0) {
            return new RodCuttingReturnObject(0, cutsTable);
        }

        if (maxProfitsTable[len] == 0) {

            for (int i = 1; i <= len; i++) {

                int profit = prices[i] + rodCuttingTopDownDp(len - i, prices, maxProfitsTable, cutsTable).getProfit();

                if (profit > maxProfitsTable[len]) {
                    maxProfitsTable[len] = profit;
                    cutsTable[len] = i;
                }

            }

        }

        return new RodCuttingReturnObject(maxProfitsTable[len], cutsTable);

    }

    private static RodCuttingAnswerObject rodCuttingBottomUpDp(int len, int[] prices) {

        int[] maxProfitsTable = new int[len + 1];
        int[] cutsTable = new int[len + 1];

        for (int i = 1; i <= len; i++) {

            for (int j = 1; j <= i; j++) {

                int profit = prices[j] + maxProfitsTable[i - j];

                if (profit > maxProfitsTable[i]) {
                    maxProfitsTable[i] = profit;
                    cutsTable[i] = j;
                }

            }

        }

        int n = len;
        List<Integer> answerTable = new ArrayList<>();

        while (n > 0) {
            answerTable.add(cutsTable[n]);
            n -= cutsTable[n];
        }

        return new RodCuttingAnswerObject(maxProfitsTable[len], answerTable);

    }

}
